src/runner/mod.rs

```diff
@@ -64,6 +64,14 @@ pub fn run_ex<DB: WriteResults + Sync>(
     let graph = Mutex::new(build_graph(ex, crates, config));
     let parked_threads = Condvar::new();
 
+    info!("uninstalling toolchains...");
+    // Clean out all the toolchains currently installed. This minimizes the
+    // amount of disk space used by the base system, letting the task execution
+    // proceed slightly faster than it would otherwise.
+    for tc in workspace.installed_toolchains()? {
+        tc.uninstall(workspace)?;
+    }
```

Comment:
This works, but it'd be better to check whether the toolchain is still needed before deleting it.

Distributed experiments in Crater are implemented on the server side by internally splitting a single experiment into multiple ones, each with [1024 crates](https://github.com/rust-lang/crater/blob/ba03080696d339981948fff0b3da3611e45fdbb9/config.toml#L15-L18) in it. This means `run_ex` is called multiple times for each "real" experiment, and this PR as is would download the toolchain multiple times on the same agent.

Reply:
How would you suggest checking if it's still needed? Do you mean just that the toolchain is not going to be re-installed ~2 lines lower? (A more fancy solution would require some kind of server "experiment done" request, right?)

Reply:
> Do you mean just that the toolchain is not going to be re-installed ~2 lines lower?

Yeah, something like this should work:

```rust
    for tc in workspace.installed_toolchains()? {
        if !ex.toolchains.contains(&tc) {
            tc.uninstall(workspace)?;
        }
    }
```

Reply:
Alright, updated.

Reply:
Uh, there is a thing I just remembered: Rustwide itself needs access to [a "main" toolchain](https://github.com/rust-lang/rustwide/blob/1b5f1742b4cd4b82d791f5c339ed376735d6842e/src/toolchain.rs#L175-L179) (stable) to be able to install tools like `rustup-toolchain-install-master` on workspace initialization.

I'm wondering how to best handle that. Hardcoding that in Crater feels wrong, maybe `Toochain::uninstall` should return an error if that toolchain is the main one?

Reply:
We already have a workspace here; maybe the correct fix is for rustwide to re-install needed toolchains on workspace initialization? That is, is that an ongoing need? Workspaces are initialized pretty rarely, so I suspect we may be able to afford it causing a redownload? (I don't know if that's true on stable).

Reply:
But returning an error seems annoying, in any case, since then all uninstalls needs to think about handling that (largely likely to be spurious) error.

Reply:
Actually, I *just* checked and Rustwide does indeed re-install the toolchain if it's missing! Resolving the concern.